runtime/vm/KeyHashTable.c

```diff
@@ -590,6 +590,102 @@ hashClassTableNextDo(J9HashTableState *walkState)
 	return (NULL == next) ? NULL : next->ramClass;
 }
 
+#if defined(J9VM_OPT_SNAPSHOTS)
+J9Class *
+hashClassTableRCPStartDo(J9ClassLoader *classLoader, J9HashTableState *walkState, UDATA flags)
+{
+	BOOLEAN skipHidden = J9_ARE_ALL_BITS_SET(flags, J9_HASH_TABLE_STATE_FLAG_SKIP_HIDDEN);
+	BOOLEAN continueToNext = FALSE;
+	KeyHashTableClassEntry *first = hashTableStartDo(classLoader->classHashTable, walkState);
+
+	if (NULL != first) {
+		if (TAG_RAM_CLASS != (first->tag & MASK_RAM_CLASS)) {
+			/* only report RAM classes */
+			continueToNext = TRUE;
+		} else {
+			if ((skipHidden && J9ROMCLASS_IS_HIDDEN(first->ramClass->romClass))
+				|| ((NULL == first->ramClass->classObject) && J9_ARE_ALL_BITS_SET(first->ramClass->classFlags, J9ClassIsFrozenFromSnapshot))
+				) {
+				continueToNext = TRUE;
+			} else {
+				continueToNext = FALSE;
+			}
+		}
+	} else {
+		continueToNext = FALSE;
+	}
+
+	while (continueToNext) {
+		first = hashTableNextDo(walkState);
+		if (NULL != first) {
+			if (TAG_RAM_CLASS != (first->tag & MASK_RAM_CLASS)) {
+				/* only report RAM classes */
+				continueToNext = TRUE;
+			} else {
+				if ((skipHidden && J9ROMCLASS_IS_HIDDEN(first->ramClass->romClass))
+					|| ((NULL == first->ramClass->classObject) && J9_ARE_ALL_BITS_SET(first->ramClass->classFlags, J9ClassIsFrozenFromSnapshot))
+				) {
+					continueToNext = TRUE;
+				} else {
+					continueToNext = FALSE;
+				}
+			}
+		} else {
+			continueToNext = FALSE;
+		}
+	}
+	walkState->flags = flags;
+	return (NULL == first) ? NULL : first->ramClass;
+}
+
+J9Class *
+hashClassTableRCPNextDo(J9HashTableState *walkState)
+{
+	BOOLEAN skipHidden = J9_ARE_ALL_BITS_SET(walkState->flags, J9_HASH_TABLE_STATE_FLAG_SKIP_HIDDEN);
+	BOOLEAN continueToNext = FALSE;
+	KeyHashTableClassEntry *next = hashTableNextDo(walkState);
+
+	if (NULL != next) {
+		if (TAG_RAM_CLASS != (next->tag & MASK_RAM_CLASS)) {
+			/* only report RAM classes */
+			continueToNext = TRUE;
+		} else {
+			if ((skipHidden && J9ROMCLASS_IS_HIDDEN(next->ramClass->romClass))
+					|| ((NULL == next->ramClass->classObject) && J9_ARE_ALL_BITS_SET(next->ramClass->classFlags, J9ClassIsFrozenFromSnapshot))
```

Comment:
- The check should be just for the flag.
- The `NULL` check on the `classObject` is not needed.
- A local variable should be used for `next->ramClass` since it is accessed multiple times.